ps: ps command implementation

devimc opened the pull request:
this patch adds the implementation for ps command, ps output will
be parsed to see what processes are running inside the given
container therefore a list of processes that belong to the container
is returned

Signed-off-by: Julio Montes <julio.montes@intel.com>

dlespiau commented:
Not sure it's a good idea to keep adding features here. We're happy to have feature-parity with 2.1 as far as hyperstart is concerned for 3.0.

devimc commented:
@dlespiau  what do you mean? should I add this new feature in master? I thought cc 3.0 was using hyperstart 0.7.0

dlespiau commented:
I hadn't noticed samuel had added docker ps support in the beta project. I guess we need that. The point is we should try to limit new features: we're not properly upstreaming - which creates a lot of technical debt - and we'd like to use our own agent at some point.

Commit:
util: add extra parameters to hyper_cmd for saving command stdout

this patch adds two extra parameters to hyper_cmd for saving command stdout
and a new function to read and save the data from command stdout fd

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
portmapping: fix portmapping according with new implementation of hyper_cmd

Signed-off-by: Julio Montes <julio.montes@intel.com>

Commit:
ps: implements ps command

this patch adds the implementation for ps command, ps output will
be parsed to see what processes are running inside the given
container therefore a list of processes that belong to the container
is returned

Signed-off-by: Julio Montes <julio.montes@intel.com>

jodh-intel commented on src/init.c:
```diff
@@ -36,6 +36,54 @@ struct hyper_pod global_pod = {
+	/**
+	 * line to process, it contains information about a process
+	 * for example:
+	 *   root 5123 2 0 09:51 ? 00:00:00 [kworker/2:0]
```
I like the example, but since this is at the top of the file, it would be good to explain that the format will vary, depending on the value of `psargs`.

devimc replied:
fixed

jodh-intel commented on src/init.c:
```diff
@@ -720,6 +768,367 @@ static int hyper_remove_container(char *json, int length)
+	char **array = NULL;
+	char *new_line = NULL;
+	int ret = -1;
+	const char* pid_str = "PID";
```
I'm wondering if it might be better to:

- Require `psargs` to specify the output format such that the first field *must be* the `pid`.
- If no `psargs` are specified, make hyperstart specify the format in the same way:
  `ps -o pid,ppid,pgrp,session,tpgid,tty,comm -e` or something).

The advantage of doing that:

- You could remove `pid_index` from `hyper_ps_data`.
- It would be easy to reject the `PSCONTAINER` command if the first field wasn't purely numeric on every line.
- You wouldn't need this function.
- The code would be safer since what happens if `psargs` is just `--no-headers` at the moment?

devimc replied:
it is a good idea indeed, but ```-o``` option will override all other options, for example:
```
$ ps -ef
UID        PID  PPID  C STIME TTY          TIME CMD
```
```
ps -o pid,ppid,pgrp,session,tpgid,tty -ef
  PID  PPID  PGRP  SESS TPGID TT
```

jodh-intel commented on src/init.c:
```diff
@@ -720,6 +768,367 @@ static int hyper_remove_container(char *json, int length)
+	/* +1 for \n */
+	size_t line_len = ps_data->line_len + 1;
+	size_t output_len = *ps_data->output_len;
+	const size_t buf_size = 1024;
```
This limit might be too low: on a test box, I've got a qemu cmdline that is `2634` bytes long.

devimc replied:
```1024``` is the initial size for the buffer,  the size will increase by 1024 in each successive ```realloc```

jodh-intel replied:
Of course. Might be a good idea to put a comment above it for clarity (or even call it `initial_buf_size` or `buf_size_increment` maybe).

devimc replied:
agree :)

dlespiau commented:
wouldn't it be possible/better to do an "exec" with the ps command instead of introducing a new command that doesn't seem to add value compared to an exec? what am I missing?

dlespiau commented:
Ah, answering to myself. We don't necessarily have ps in the container rootfs. This adds a new dependency for the guest OS though.

devimc commented:
@dlespiau yes, is not a guarantee that container rootfs will have the ps command

jodh-intel commented on src/init.c:
```diff
@@ -720,6 +768,367 @@ static int hyper_remove_container(char *json, int length)
+ *
+ * \return pid index on success, else -1
+ */
+static int hyper_get_pid_index(const char *line)
```
Missing a check on `line` being `NULL`.

devimc replied:
fixing, thanks

jodh-intel commented on src/init.c:
```diff
@@ -720,6 +768,367 @@ static int hyper_remove_container(char *json, int length)
+ *
+ * \return true on success, else false
+ */
+static bool hyper_append_ps_line(struct hyper_ps_data *ps_data)
```
Check `ps_data` against `NULL` before dereferencing.

devimc replied:
fixing

jodh-intel commented on src/init.c:
```diff
@@ -720,6 +768,367 @@ static int hyper_remove_container(char *json, int length)
+ *
+ * \return true on success, else false
+ */
+static bool hyper_process_ps_line(struct hyper_ps_data *ps_data)
```
Missing check on `ps_data`.

devimc replied:
fixing

jodh-intel commented on src/init.c:
```diff
@@ -720,6 +768,367 @@ static int hyper_remove_container(char *json, int length)
+	array = hyper_line_to_array(new_process_line, &array_size, ps_data->pid_index+1);
+	if (array == NULL) {
+		free_if_set(new_process_line);
+		fprintf(stderr, "can not convert line to array: %s\n", new_process_line);
```
The `fprintf()` needs to be before the `free_if_set()`.

devimc replied:
good catch!
thanks

jodh-intel commented on src/init.c:
```diff
@@ -720,6 +768,367 @@ static int hyper_remove_container(char *json, int length)
+	}
+
+	/* processes less than container pid are not part of the container */
+	if (pid < ps_data->container_pid) {
```
Is this test reliable I wonder? The `container_pid` is in a separate namespace so might be `1` inside the container and `999` outside it in the VM context.

What if the container creates lots of processes and reaches the limit of a `pid_t`. Will the next pid be guaranteed to be `> container_pid` (`>999` in the VM context in this example)?

devimc replied:
you're right it is not a valid condition once limit is reached it is restarted

jodh-intel commented on src/init.c:
```diff
@@ -720,6 +768,367 @@ static int hyper_remove_container(char *json, int length)
+		 * this validation is to avoid command injections
+		 */
+		if (!strstr(arg, "$") && !strstr(arg, "`")) {
+			bytes_copied += snprintf(command+bytes_copied,
```
`snprintf()` can return `-1` on error which isn't handled here.

devimc replied:
fixed

jodh-intel commented on src/init.c:
```diff
@@ -720,6 +768,367 @@ static int hyper_remove_container(char *json, int length)
+ *
+ * \return 0 on success, else -1
+ */
+static int hyper_ps_container(char *json, int length, uint32_t *datalen, uint8_t **data)
```
Args need checks.

devimc replied:
fixed

jodh-intel commented on src/util.c:
```diff
@@ -775,15 +775,91 @@ void hyper_shutdown()
+	size_t bufsize = 512;
+	uint8_t *buffer = NULL;
+
+	buffer = calloc(bufsize, sizeof(*buffer));
```
What if `calloc()` fails?

devimc replied:
adding validation, thanks

jodh-intel commented on src/util.c:
```diff
@@ -775,15 +775,91 @@ void hyper_shutdown()
+	buffer = calloc(bufsize, sizeof(*buffer));
+
+	do {
+		bytes_read = read(stdout_pipe,
```
The tests in this block do seem safe, but I'd still suggest checking if `bytes_read < 0` and handling that immediately rather than implicitly.

jodh-intel commented on src/util.c:
```diff
@@ -775,15 +775,91 @@ void hyper_shutdown()
+		}
+		if (total_bytes_read >= bufsize-1) {
+			uint8_t *tmp = NULL;
+			bufsize+=bufsize;
```
Just a check: do you intend to double the size (I suspect so), or do you want to add `512`?

devimc replied:
I want to double the size

jodh-intel commented on src/util.c:
```diff
@@ -775,15 +775,91 @@ void hyper_shutdown()
+ *
+ * \return 0 on success, else -1
+ */
+static int hyper_read_cmd_stdout(int stdout_pipe, uint8_t **output, size_t *size)
```
Missing checks on `stdout_pipe` and `output` params.

devimc replied:
fixing

amshinde commented on src/util.c:
```diff
@@ -775,15 +775,111 @@ void hyper_shutdown()
+
+	if (buffer == NULL) {
+		/* try once again */
+		buffer = calloc(bufsize, sizeof(*buffer));
```
If calloc fails, it usually indicates OOM. No point trying again.

devimc replied:
I prefer to keep it, if OOM occurs hyperstart will fail any way

amshinde commented on src/util.c:
```diff
@@ -775,15 +775,111 @@ void hyper_shutdown()
+
+		if (bytes_read > 0) {
+			total_bytes_read += bytes_read;
+			buffer[total_bytes_read] = '\0';
```
This assignment can be moved out of the loop I think to perform it just once

devimc replied:
yes, thank you!

amshinde commented on src/util.c:
```diff
@@ -775,15 +775,111 @@ void hyper_shutdown()
+			buffer[total_bytes_read] = '\0';
+		}
+
+		if (total_bytes_read >= bufsize-1) {
```
Does this have to be > rather than >= check here?

devimc replied:
agree

egernst commented:
can we close this?  :)
